code/go/0chain.net/smartcontract/dbs/event/blobber.go

```diff
@@ -1,28 +1,32 @@
 package event
 
 import (
-	"0chain.net/smartcontract/dbs"
 	"errors"
 	"fmt"
+	"time"
+
+	"0chain.net/smartcontract/dbs"
+
+	"github.com/guregu/null"
 
 	"gorm.io/gorm"
 )
 
 type Blobber struct {
 	gorm.Model
 	BlobberID string `json:"id" gorm:"uniqueIndex"`
-	BaseURL   string `json:"url"`
+	BaseURL   string `json:"url" gorm:"uniqueIndex"`
 
 	// geolocation
 	Latitude  float64 `json:"latitude"`
 	Longitude float64 `json:"longitude"`
 
 	// terms
-	ReadPrice               int64   `json:"read_price"`
-	WritePrice              int64   `json:"write_price"`
-	MinLockDemand           float64 `json:"min_lock_demand"`
-	MaxOfferDuration        string  `json:"max_offer_duration"`
-	ChallengeCompletionTime string  `json:"challenge_completion_time"`
+	ReadPrice               int64         `json:"read_price"`
+	WritePrice              int64         `json:"write_price"`
+	MinLockDemand           float64       `json:"min_lock_demand"`
+	MaxOfferDuration        time.Duration `json:"max_offer_duration"`
+	ChallengeCompletionTime time.Duration `json:"challenge_completion_time"`
```

Comment:
Piers would you mind if I create separate issue for benchmark tests and delegate it to someone? I'm afraid that it will take some additional time and can be the good first issue to dig a little into smartcontracts?

Reply:
I think we need to add a benchmark test to the GitHub actions to make sure there are no failing tests, before a PR can be merged.

MaxOfferDuration and ChallengCompletionTime were working before, so what about changing back to a string and having a separate issue to convert MaxOfferDuration and ChallengCompletionTime to a custom type.

In practice leaving broken benchmark tests means I will have to fix them in a couple of weeks' time, with everyone else's messed up benchmark tests. when I am asked to present all the latest benchmark times.

Reply:
Yes, we need a github actions to run benchmark test for PRs

Reply:
@Sriep I was able to run tests from scratch, looks like something is wrong with your setup
```
Results
benchmark suite storage
commit_connection                          2.219758ms OK
storage.add_blobber                        0.823649ms OK
storage.add_curator                        0.686613ms OK
storage.add_free_storage_assigner          0.329697ms OK
storage.add_validator                      0.365569ms FAILED - add_validator_failed: Failed to get validator list.value not present
storage.blobber_block_rewards              2.572570ms OK
storage.blobber_health_check               0.518838ms OK
storage.cancel_allocation                  0.664914ms FAILED - alloc_cancel_failed: trying to cancel expired allocation
storage.challenge_response                 0.568006ms FAILED - verify_challenge: could not get the blobber challenge d46458063f43eb4aeb4adf1946d123908ef63143858abb24376d42b5761bf577: value not present
storage.collect_reward                     0.415690ms OK
storage.curator_transfer_allocation        1.263867ms OK
storage.finalize_allocation                4.787383ms OK
storage.free_allocation_request            2.954802ms FAILED - free_allocation_failed: creating new allocation: allocation_creation_failed: Blobbers provided are not enough to honour the allocation
storage.free_update_allocation             3.751701ms OK
storage.generate_challenge                 0.045721ms OK
storage.new_allocation_request_preferred   0.337830ms FAILED - allocation_creation_failed: Not enough provided blobbers found in mpt
storage.new_allocation_request_random      0.458543ms FAILED - allocation_creation_failed: Blobbers provided are not enough to honour the allocation
storage.new_read_pool                      0.240293ms OK
storage.read_pool_lock                     0.493596ms OK
storage.read_pool_unlock                   0.233326ms OK
storage.read_redeem                        2.353349ms OK
storage.remove_curator                     0.432118ms OK
storage.stake_pool_lock                    0.505390ms OK
storage.stake_pool_unlock                  0.649737ms OK
storage.update_allocation_request          1.145969ms FAILED - allocation_updating_failed: can't update expired allocation
storage.update_blobber_settings            1.571320ms OK
storage.update_settings                    0.622626ms OK
storage.write_pool_lock                    0.620023ms OK
storage.write_pool_unlock                  0.872665ms OK
benchmark suite storage_rest
storage_rest.GetAllocBlobber               0.184704ms FAILED - allocation_creation_failed: Too many blobbers selected                                                                                                                      max available 0
storage_rest.GetValidator                  0.233089ms FAILED - internal_error: can't get validator: error retriving Validation node with ID 365c66a8f657eb0270fd14b76105a774bcc161c091b4acce8cda5a17396c338b; error: record not found
storage_rest.allocation                    1.152139ms OK
storage_rest.allocation_min_lock           0.516595ms OK
storage_rest.allocations                   0.932109ms OK
storage_rest.countreadmarkers              0.461424ms OK
storage_rest.getBlobber                    0.182191ms OK
storage_rest.getChallengePoolStat          0.078757ms OK
storage_rest.getConfig                     0.124900ms OK
storage_rest.getReadPoolAllocBlobberStat   0.041429ms OK
storage_rest.getReadPoolStat               0.043462ms OK
storage_rest.getStakePoolStat              0.927920ms OK
storage_rest.getUserStakePoolStat          0.227408ms OK
storage_rest.getWriteMarkers               0.129823ms OK
storage_rest.getWritePoolAllocBlobberStat  0.093055ms OK
storage_rest.getWritePoolStat              0.053308ms OK
storage_rest.get_mpt_key.sc_config         0.000409ms OK
storage_rest.getblobbers                   0.305681ms OK
storage_rest.getchallenge                  0.338370ms FAILED - resource_not_found: can't get blobber challenge: value not present
storage_rest.latestreadmarker              0.057786ms OK
storage_rest.openchallenges                0.144500ms OK
storage_rest.readmarkers                   0.166393ms OK
```

Reply:
The benchmark code seems to have been recently broken, panicking [here](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/zcnsc/benchmark_setup.go#L38). Probably due to recent PRs that changed the zcn GlobalNode. This indicates why a GitHub action to run benchmark tests would be a nice idea.

Reply:
@Sriep maybe something is wrong with your db setup? I don't recommend to change int db field to string. Benchmarks are working locally, I can't reproduce your error

Reply:
please recreate scheme